### PR TITLE
Remove default impl of Collect::current_span

### PR DESCRIPTION
**Pull request by oleschoenburg**
## Motivation

This removes the default implementation of `tracing_core::Collect::current_span`, which implements part one of #1262.

## Solution
1. Removes the default implementation
2. Makes `tracing_core::span::Current::unknown` public so that implementers of `Collect` can use it if they didn't implement `current_span()` before.
3. Copy the old default implementation to all implementers of `Collect`

### CODE REVIEW

**Comment by dignati** [comment]
We could also add a re-export of `tracing_core::span::Current` as `tracing::span::Current` which would look a bit nicer when implementing `Collect`.

**Comment by dignati** [comment]
I keep forgetting to run `cargo check --all --bins --tests --benches` so there are a bunch of places still left that now need to implement `current_span`.

File: tracing/tests/support/collector.rs
Comment (davidbarsky) [inline tracing/tests/support/collector.rs:526]: Why's this one todo'd, whereas all the others are unknown?

File: examples/examples/sloggish/sloggish_collector.rs
Comment (hawkw) [inline examples/examples/sloggish/sloggish_collector.rs:271]: in this example, the collector *does* know the current span's ID: https://github.com/tokio-rs/tracing/blob/f867e83dbaaa9ae73d7ee26a6766bb9b49730d06/examples/examples/sloggish/sloggish_collector.rs#L58

so, maybe this should return that span? this would require a slightly larger change, modifying [the `Span` struct](https://github.com/tokio-rs/tracing/blob/f867e83dbaaa9ae73d7ee26a6766bb9b49730d06/examples/examples/sloggish/sloggish_collector.rs#L66-L69) to also include a `&'static Metadata<'static>` that's stored in `new_span`, but then we could write something like this:
```suggestion
        if let Some(id) = self.current.id() {
            let metadata = self.spans
                .lock()
                .unwrap()
                .get(id)
                .unwrap_or_else(|| panic!("no metadata stored for span with ID {:?}", id))
                .metadata;
            return Current::new(id, metadata);
        }
        
        Current::none()
```

File: tracing-core/src/collect.rs
Comment (hawkw) [inline tracing-core/src/collect.rs:410]: I think we need to add a link reference for `unknown` here:
```suggestion
    /// [`Current::none`]: super::span::Current::none
    /// [`Current::unknown`]: super::span::Current::unknown
```

File: tracing-core/src/dispatch.rs
Comment (hawkw) [inline tracing-core/src/dispatch.rs:862]: nit, this doesn't really matter: i think it _might_ be more correct for the no-op collector to return `none` rather than `unknown` — the intended semantics is that all spans are _disabled_ when there's no collector, so we are never in a span. with that said, it doesn't really make a difference in practice.

File: tracing/tests/support/collector.rs
Comment (hawkw) [reply tracing/tests/support/collector.rs:526]: the test collector probably _should_ actually implement this, it knows the ID of the current span:
https://github.com/tokio-rs/tracing/blob/f867e83dbaaa9ae73d7ee26a6766bb9b49730d06/tracing/tests/support/collector.rs#L44
and it can look up spans by ID:
https://github.com/tokio-rs/tracing/blob/f867e83dbaaa9ae73d7ee26a6766bb9b49730d06/tracing/tests/support/collector.rs#L42

similar to the `sloggish` example collector, we would need to add the metadata to the stored span data to implement this properly. in this case, the stored span data is the `SpanState` struct:
https://github.com/tokio-rs/tracing/blob/f867e83dbaaa9ae73d7ee26a6766bb9b49730d06/tracing/tests/support/collector.rs#L36-L39

we could pretty easily add a `&'static Metadata<'static>` to that, allowing us to actually return the current span.

i think this is somewhat important; even though none of the tests currently use the mock collector's `current_span` method, it should probably implement the behavior correctly so that we're not surprised when writing future tests that expect it to work...

**Review by hawkw — CHANGES_REQUESTED** [review]
this feels like a good start. i think there's a few places where we probably ought to add real `current_span` implementations instead of just returning `unknown`. once that's taken care of, I'll be happy to merge this!

**Comment by dignati** [comment]
Sorry, I should have marked this as WIP earlier. I'll try to implement current_span properly where it makes sense but it might take me a while. I don't want to take up too much of your resources or block the release of 0.2 so feel free work on it yourself in the meantime.

**Comment by dignati** [comment]
Okay, I think I implemented all suggestions. From my side there is only one open question: https://github.com/tokio-rs/tracing/pull/1268#issuecomment-786198583

**Comment by davidbarsky** [comment]
@dignati Sorry for the delay in responding.

> We could also add a re-export of tracing_core::span::Current as tracing::span::Current which would look a bit nicer when implementing Collect.

I think we can keep this as-is for now and not worry about it. If you could rebase again the master branch (or I can do it for you!), then I think we're clear to merge this in.

**Comment by dignati** [comment]
No worries! I rebased without any conflicts.

**Comment by oleschoenburg** [comment]
Is this PR still relevant? I can rebase again if needed, otherwise feel free to close this.

**Review by hawkw — APPROVED** [review]
okay, this looks good to me --- thanks for the ping!

**Comment by hawkw** [comment]
Ah, it looks like, post-rebase, the build is failing now because new `Collect` impls were added on master --- for `Box<dyn Collect + ...>` and `Arc<dyn Collect + ...>`. The new impls are here: https://github.com/tokio-rs/tracing/blob/44d355800f8e7da2dfc601289d344b1967c702be/tracing-core/src/collect.rs#L559-L691

We should add `current_span` methods to those new `Collect` impls, and then CI should pass again.

**Review by davidbarsky — APPROVED** [review]

**Comment by davidbarsky** [comment]
Thanks for working on this and sorry for the delay in merging!

**Pull request merged**
